File: python/tests/test_dupire.py
"""Dupire local vol: flat-surface identity, clamping, skew/smile behaviour."""

from __future__ import annotations

import math

import numpy as np
import pytest

from localvol import CAP, FLOOR, DupireLocalVol, ImpliedVolSurface


def test_flat_surface_localvol_equals_implied(flat_surface):
    """Property-style golden anchor: sigma_loc == 20% to 1e-6 everywhere."""
    lv = DupireLocalVol(flat_surface)
    for k in np.linspace(-0.45, 0.45, 9):
        for t in [0.05, 0.25, 0.6, 1.0, 1.7, 2.4]:
            assert float(lv.vol(float(k), float(t))) == pytest.approx(0.20, abs=1e-6)
    assert lv.floor_count == 0 and lv.cap_count == 0


def test_bundled_surface_sane_interior(bundled_localvol):
    """Skewed surface: local vol finite, in band, decreasing across the skew."""
    lv = bundled_localvol
    ks = np.linspace(-0.25, 0.25, 11)
    vols = np.asarray(lv.vol(ks, 1.0))
    assert np.all(np.isfinite(vols))
    assert np.all((vols >= FLOOR) & (vols <= CAP))
    # negative-rho SSVI: local vol decreases from put wing to call wing
    assert vols[0] > vols[-1]
    # local skew is roughly twice the implied skew — at least steeper:
    srf = lv.surface
    imp = np.asarray(srf.implied_vol(ks, 1.0))
    assert (vols[0] - vols[-1]) > (imp[0] - imp[-1])


def test_vectorised_matches_scalar(bundled_localvol):
    ks = np.array([-0.3, -0.1, 0.0, 0.15, 0.3])
    vec = np.asarray(bundled_localvol.vol(ks, 0.8))
    sca = np.array([float(bundled_localvol.vol(float(k), 0.8)) for k in ks])
    np.testing.assert_allclose(vec, sca, atol=1e-15)


def test_floor_engaged_on_extreme_skew():
    """A calendar-flat (dw/dT ~ 0) steep surface must clamp, not crash."""
    ks = np.linspace(-0.5, 0.5, 11)
    ts = np.array([0.5, 1.0])
    # nearly constant total variance in T -> numerator ~ 0 -> floor engages
    v1 = 0.2 + 0.5 * np.abs(ks)
    v2 = np.sqrt((v1**2 * 0.5 + 1e-9) / 1.0)
    srf = ImpliedVolSurface(ks, ts, np.vstack([v1, v2]))
    lv = DupireLocalVol(srf)
    with np.errstate(all="ignore"):
        vols = np.asarray(lv.vol(ks, 0.75))
    assert np.all(vols >= FLOOR - 1e-15)
    assert lv.floor_count > 0
    assert "floor" in lv.violation_report
    lv.reset_counters()
    assert lv.floor_count == 0 and lv.cap_count == 0


def test_cap_engaged_on_butterfly_violating_wing():
    """Over-steep smile (butterfly arb): denominator <= 0 -> capped at 500%."""
    ks = np.linspace(-0.5, 0.5, 21)
    ts = np.array([0.5, 1.0])
    v = 0.10 + 1.0 * np.abs(ks)  # V-shaped smile: wings violate |dw/dk| bound
    srf = ImpliedVolSurface(ks, ts, np.vstack([v, v * 1.05]))
    lv = DupireLocalVol(srf)
    with np.errstate(all="ignore"):
        vols = np.asarray(lv.vol(np.linspace(-0.45, 0.45, 41), 0.75))
    assert np.all(vols <= CAP + 1e-15)
    assert lv.cap_count > 0


def test_fx_symmetric_smile():
    """FX-style symmetric smile: local vol symmetric-ish and smile-shaped."""
    ks = np.linspace(-0.3, 0.3, 13)
    ts = np.array([0.25, 0.5, 1.0])
    vols = np.array([[0.10 + 0.15 * k * k for k in ks] for _ in ts])
    srf = ImpliedVolSurface(ks, ts, vols * np.array([1.0, 1.0, 1.0])[:, None])
    lv = DupireLocalVol(srf)
    v_pos = float(lv.vol(0.15, 0.5))
    v_neg = float(lv.vol(-0.15, 0.5))
    v_atm = float(lv.vol(0.0, 0.5))
    assert v_pos == pytest.approx(v_neg, rel=0.15)  # near-symmetric
    assert v_pos > v_atm and v_neg > v_atm          # smile shape survives


def test_validation():
    with pytest.raises(ValueError):
        DupireLocalVol("not a surface")  # type: ignore[arg-type]
    ks = np.array([-0.1, 0.0, 0.1])
    srf = ImpliedVolSurface(ks, np.array([0.5, 1.0]), np.full((2, 3), 0.2))
    lv = DupireLocalVol(srf)
    with pytest.raises(ValueError):
        lv.vol(0.0, -1.0)
    with pytest.raises(ValueError):
        lv.vol(float("inf"), 1.0)


def test_no_clamp_inside_quoted_box_and_continuous_at_wings(bundled_surface):
    """CRIT-1 regression: the wing-clamped stencil must never cap inside the
    quoted box, and local vol must be continuous across the last quoted
    strike (before the fix vol(0.4995..0.5005, 1) was capped at 500%)."""
    lv = DupireLocalVol(bundled_surface)
    ks = np.arange(-0.5, 0.5 + 1e-9, 0.005)
    for t in [0.1, 0.5, 0.75, 1.0, 2.0, 3.0]:
        vols = np.asarray(lv.vol(ks, t))
        assert np.all((vols > FLOOR) & (vols < CAP)), f"T={t}: {vols.min()}..{vols.max()}"
        for edge in (bundled_surface.k_min, bundled_surface.k_max):
            lo = float(lv.vol(edge - 1e-4, t))
            hi = float(lv.vol(edge + 1e-4, t))
            assert abs(lo - hi) < 1e-3, f"T={t} k={edge}: {lo} vs {hi}"
            # constant in k beyond k_max - DK (clamped stencil)
            far = float(lv.vol(edge + math.copysign(1.0, edge), t))
            assert far == pytest.approx(hi if edge > 0 else lo, abs=1e-15)
    assert lv.floor_count == 0 and lv.cap_count == 0
    # Reference point at the wing for T = 1: 0.197 (was 5.0 before the fix).
    assert float(lv.vol(0.5, 1.0)) == pytest.approx(0.19703445, abs=1e-6)


def test_stencil_clamp_only_when_box_wider_than_2dk():
    """A degenerate box narrower than 2 DK is left unclamped (no crash)."""
    ks = np.array([-0.0005, 0.0, 0.0005])
    ts = np.array([0.5, 1.0])
    srf = ImpliedVolSurface(ks, ts, np.full((2, 3), 0.2))
    lv = DupireLocalVol(srf)
    assert float(lv.vol(0.0, 0.75)) == pytest.approx(0.2, abs=1e-6)
    assert float(lv.vol(3.0, 0.75)) == pytest.approx(0.2, abs=1e-6)


def test_term_structure_local_vol_reprices_forward_variance():
    """Flat-in-k surface with iv = {0.15, 0.20, 0.25} at T = {0.25, 0.5, 1}:
    local variance is the piecewise-constant forward variance."""
    ks = np.linspace(-0.5, 0.5, 11)
    ts = np.array([0.25, 0.5, 1.0])
    vols = np.array([[0.15] * 11, [0.20] * 11, [0.25] * 11])
    srf = ImpliedVolSurface(ks, ts, vols)
    lv = DupireLocalVol(srf)
    # (0.20^2*0.5 - 0.15^2*0.25) / 0.25 = 0.0575 -> 0.239792
    assert float(lv.vol(0.0, 0.4)) == pytest.approx(math.sqrt(0.0575), abs=1e-6)
    assert float(lv.vol(0.3, 0.4)) == pytest.approx(math.sqrt(0.0575), abs=1e-6)
    # (0.25^2*1 - 0.20^2*0.5) / 0.5 = 0.085 -> 0.291548
    assert float(lv.vol(0.0, 0.75)) == pytest.approx(math.sqrt(0.085), abs=1e-6)
    # below the first pillar: flat forward variance = first-pillar vol
    assert float(lv.vol(0.0, 0.1)) == pytest.approx(0.15, abs=1e-6)
    # beyond the last pillar: last interval's slope continues
    assert float(lv.vol(0.0, 2.0)) == pytest.approx(math.sqrt(0.085), abs=1e-6)
    assert lv.floor_count == 0 and lv.cap_count == 0


def test_term_structure_pde_round_trip_within_3bp():
    """End to end: pillar kinks + t_mid freezing + time interpolation."""
    from localvol import Market, implied_vol, price_european_pde

    ks = np.linspace(-0.5, 0.5, 11)
    ts = np.array([0.25, 0.5, 1.0])
    vols = np.array([[0.15] * 11, [0.20] * 11, [0.25] * 11])
    srf = ImpliedVolSurface(ks, ts, vols)
    lv = DupireLocalVol(srf)
    mkt = Market(100.0, 0.03, 0.01)
    for t in (0.75, 1.0):
        iv_in = float(srf.implied_vol(0.0, t))
        price = price_european_pde(mkt, 100.0, t, lv.vol, num_space=200, num_time=200, sigma_ref=iv_in)
        iv_out = implied_vol(price, 100.0, 100.0, 0.03, 0.01, t, True)
        assert abs(iv_out - iv_in) * 1e4 < 3.0, f"T={t}: {(iv_out - iv_in) * 1e4:.2f} bp"


def test_t0_equals_short_time_limit(bundled_localvol):
    """MAJ-1 regression: vol(k, 0) is the Berestycki-Busca-Florent limit
    sigma_imp / (1 - k sigma_imp'/sigma_imp), continuous with T -> 0+
    (before the fix vol(-0.3, 0) = 0.309 vs vol(-0.3, 1e-6) = 0.452)."""
    lv = bundled_localvol
    srf = lv.surface
    for k in (-0.3, -0.1, 0.0, 0.2, 0.45):
        v0 = float(lv.vol(k, 0.0))
        assert abs(v0 - float(lv.vol(k, 1e-6))) < 1e-3
        assert abs(v0 - float(lv.vol(k, 1e-4))) < 1e-3
        # independent BBF evaluation from the short-end implied slice
        s = float(srf.implied_vol(k, 0.0))
        sp = (float(srf.implied_vol(k + 1e-3, 0.0)) - float(srf.implied_vol(k - 1e-3, 0.0))) / 2e-3
        assert v0 == pytest.approx(s / (1.0 - k * sp / s), abs=1e-12)
    assert float(lv.vol(-0.3, 0.0)) == pytest.approx(0.45240187, abs=1e-6)
    assert float(lv.vol(0.0, 0.0)) == pytest.approx(0.2, abs=1e-12)  # ATM: s' cancels


def test_t0_limit_caps_on_butterfly_violating_slope():
    """1 - k s'/s <= 0 at T = 0 must cap (and count), never divide by zero."""
    ks = np.linspace(-0.5, 0.5, 21)
    # Strongly convex parabola: at |k| = 0.5, k s'/s = 0.5 * 4 / 1.05 = 1.9 > 1.
    # (A linear or mildly convex smile never violates it: s - k s' is the
    # tangent's intercept at k = 0, which convexity keeps above s(0).)
    v = 0.05 + 4.0 * ks * ks
    srf = ImpliedVolSurface(ks, np.array([0.5, 1.0]), np.vstack([v, v * 1.05]))
    lv = DupireLocalVol(srf)
    vols = np.asarray(lv.vol(ks, 0.0))
    assert np.all(vols <= CAP + 1e-15) and np.all(vols >= FLOOR - 1e-15)
    assert lv.cap_count > 0
    assert float(lv.vol(0.5, 0.0)) == CAP
    assert float(lv.vol(0.0, 0.0)) == pytest.approx(0.05, abs=1e-9)

File: cpp/tests/test_dupire.cpp
// Dupire local vol: flat-surface identity, clamp counters, validation.

#include <gtest/gtest.h>

#include <cmath>
#include <stdexcept>
#include <string>
#include <thread>
#include <type_traits>
#include <vector>

#include "localvol/dupire.hpp"
#include "localvol/surface.hpp"

using localvol::DupireLocalVol;
using localvol::ImpliedVolSurface;

namespace {
const std::string kDataDir = LOCALVOL_DATA_DIR;
}

TEST(Dupire, FlatSurfaceGivesFlatLocalVolEverywhere) {
    // Property-style grid loop: a flat implied surface must collapse the
    // Gatheral formula to sigma_loc == implied vol identically (to 1e-6).
    const ImpliedVolSurface srf = ImpliedVolSurface::from_csv(kDataDir + "/flat_surface.csv");
    const DupireLocalVol lv(srf);
    for (double k : {-0.45, -0.2, 0.0, 0.15, 0.45}) {
        for (double t : {0.1, 0.25, 0.7, 1.0, 1.9, 2.5}) {
            EXPECT_NEAR(lv.vol(k, t), 0.20, 1e-6) << "k=" << k << " T=" << t;
        }
    }
    EXPECT_EQ(lv.floor_count(), 0);
    EXPECT_EQ(lv.cap_count(), 0);
}

TEST(Dupire, BundledSurfaceSkewShape) {
    // Equity-style skew: local vol on the put wing exceeds the call wing.
    const ImpliedVolSurface srf = ImpliedVolSurface::from_csv(kDataDir + "/implied_surface.csv");
    const DupireLocalVol lv(srf);
    const double put_wing = lv.vol(-0.25, 1.0);
    const double atm = lv.vol(0.0, 1.0);
    const double call_wing = lv.vol(0.25, 1.0);
    EXPECT_GT(put_wing, atm);
    EXPECT_LT(call_wing, atm);
    for (double v : {put_wing, atm, call_wing}) {
        EXPECT_GE(v, DupireLocalVol::kFloor);
        EXPECT_LE(v, DupireLocalVol::kCap);
    }
}

TEST(Dupire, FloorEngagesOnCalendarArbitrageAndIsCounted) {
    // Total variance *decreasing* in T at k=0 -> dw/dT <= 0 -> floor.
    std::vector<double> k = {-0.3, 0.0, 0.3};
    std::vector<double> t = {0.5, 1.0};
    std::vector<std::vector<double>> v = {
        {0.30, 0.30, 0.30},
        {0.10, 0.10, 0.10},  // w drops from 0.045 to 0.01
    };
    const ImpliedVolSurface srf(k, t, v);
    EXPECT_GT(srf.calendar_violations(), 0);
    const DupireLocalVol lv(srf);
    EXPECT_DOUBLE_EQ(lv.vol(0.0, 0.75), DupireLocalVol::kFloor);
    EXPECT_GE(lv.floor_count(), 1);
    lv.reset_counters();
    EXPECT_EQ(lv.floor_count(), 0);
    EXPECT_EQ(lv.cap_count(), 0);
}

TEST(Dupire, ReportStringMentionsCounts) {
    const ImpliedVolSurface srf = ImpliedVolSurface::from_csv(kDataDir + "/flat_surface.csv");
    const DupireLocalVol lv(srf);
    (void)lv.vol(0.0, 1.0);
    const std::string rep = lv.violation_report();
    EXPECT_NE(rep.find("floor"), std::string::npos);
    EXPECT_NE(rep.find("cap"), std::string::npos);
}

TEST(Dupire, RejectsInvalidQueries) {
    const ImpliedVolSurface srf = ImpliedVolSurface::from_csv(kDataDir + "/flat_surface.csv");
    const DupireLocalVol lv(srf);
    EXPECT_THROW(lv.vol(0.0, -0.5), std::invalid_argument);
    EXPECT_THROW(lv.vol(std::nan(""), 1.0), std::invalid_argument);
    EXPECT_THROW(lv.local_variance(0.0, std::nan("")), std::invalid_argument);
}

TEST(Dupire, TemporaryBindingIsACompileError) {
    // MAJ-2: `DupireLocalVol lv(ImpliedVolSurface::from_csv(path));` would
    // dangle; the rvalue overload is deleted so the code does not compile.
    static_assert(!std::is_constructible_v<DupireLocalVol, ImpliedVolSurface>,
                  "DupireLocalVol must not accept a temporary surface");
    static_assert(std::is_constructible_v<DupireLocalVol, const ImpliedVolSurface&>,
                  "DupireLocalVol must accept an lvalue surface");
    static_assert(!std::is_copy_constructible_v<DupireLocalVol>, "shared by reference");
}

TEST(Dupire, NoCapInsideQuotedBoxAndContinuousAtWings) {
    // CRIT-1: before the stencil clamp vol(0.4995..0.5005, 1) was 5.0.
    const ImpliedVolSurface srf = ImpliedVolSurface::from_csv(kDataDir + "/implied_surface.csv");
    const DupireLocalVol lv(srf);
    for (double t : {0.1, 0.5, 0.75, 1.0, 2.0, 3.0}) {
        for (int i = 0; i <= 200; ++i) {
            const double k = -0.5 + 0.005 * i;
            const double v = lv.vol(k, t);
            EXPECT_GT(v, DupireLocalVol::kFloor) << "k=" << k << " T=" << t;
            EXPECT_LT(v, DupireLocalVol::kCap) << "k=" << k << " T=" << t;
        }
        for (double edge : {srf.k_min(), srf.k_max()}) {
            const double lo = lv.vol(edge - 1e-4, t);
            const double hi = lv.vol(edge + 1e-4, t);
            EXPECT_LT(std::abs(lo - hi), 1e-3) << "T=" << t << " k=" << edge;
            // constant in k beyond k_max - DK (clamped stencil)
            const double far = lv.vol(edge + (edge > 0 ? 1.0 : -1.0), t);
            EXPECT_DOUBLE_EQ(far, edge > 0 ? hi : lo);
        }
    }
    EXPECT_EQ(lv.floor_count(), 0);
    EXPECT_EQ(lv.cap_count(), 0);
    EXPECT_NEAR(lv.vol(0.5, 1.0), 0.19703445, 1e-6);
}

TEST(Dupire, TermStructureLocalVolIsForwardVariance) {
    // Flat in k, iv = {0.15, 0.20, 0.25} at T = {0.25, 0.5, 1}.
    std::vector<double> ks;
    for (int i = 0; i < 11; ++i) ks.push_back(-0.5 + 0.1 * i);
    const std::vector<double> ts = {0.25, 0.5, 1.0};
    const std::vector<std::vector<double>> v = {
        std::vector<double>(11, 0.15), std::vector<double>(11, 0.20), std::vector<double>(11, 0.25)};
    const ImpliedVolSurface srf(ks, ts, v);
    const DupireLocalVol lv(srf);
    EXPECT_NEAR(lv.vol(0.0, 0.4), std::sqrt(0.0575), 1e-6);   // (0.02 - 0.005625)/0.25
    EXPECT_NEAR(lv.vol(0.3, 0.4), std::sqrt(0.0575), 1e-6);
    EXPECT_NEAR(lv.vol(0.0, 0.75), std::sqrt(0.085), 1e-6);   // (0.0625 - 0.02)/0.5
    EXPECT_NEAR(lv.vol(0.0, 0.1), 0.15, 1e-6);                // below first pillar
    EXPECT_NEAR(lv.vol(0.0, 2.0), std::sqrt(0.085), 1e-6);    // beyond last pillar
    EXPECT_EQ(lv.floor_count(), 0);
    EXPECT_EQ(lv.cap_count(), 0);
}

TEST(Dupire, ZeroExpiryIsShortTimeLimit) {
    // MAJ-1: before the fix vol(-0.3, 0) = 0.309 vs vol(-0.3, 1e-6) = 0.452.
    const ImpliedVolSurface srf = ImpliedVolSurface::from_csv(kDataDir + "/implied_surface.csv");
    const DupireLocalVol lv(srf);
    for (double k : {-0.3, -0.1, 0.0, 0.2, 0.45}) {
        const double v0 = lv.vol(k, 0.0);
        EXPECT_LT(std::abs(v0 - lv.vol(k, 1e-6)), 1e-3) << "k=" << k;
        EXPECT_LT(std::abs(v0 - lv.vol(k, 1e-4)), 1e-3) << "k=" << k;
        // independent BBF evaluation from the short-end implied slice
        const double s = srf.implied_vol(k, 0.0);
        const double sp = (srf.implied_vol(k + 1e-3, 0.0) - srf.implied_vol(k - 1e-3, 0.0)) / 2e-3;
        EXPECT_NEAR(v0, s / (1.0 - k * sp / s), 1e-12) << "k=" << k;
    }
    EXPECT_NEAR(lv.vol(-0.3, 0.0), 0.45240187, 1e-6);
    EXPECT_NEAR(lv.vol(0.0, 0.0), 0.2, 1e-12);
}

TEST(Dupire, ZeroExpiryCapsOnStronglyConvexSmile) {
    // At |k| = 0.5, k s'/s = 0.5 * 4 / 1.05 = 1.9 > 1 -> denominator <= 0 -> cap.
    std::vector<double> ks;
    std::vector<double> v1;
    for (int i = 0; i < 21; ++i) {
        const double k = -0.5 + 0.05 * i;
        ks.push_back(k);
        v1.push_back(0.05 + 4.0 * k * k);
    }
    std::vector<double> v2;
    for (double v : v1) v2.push_back(v * 1.05);
    const ImpliedVolSurface srf(ks, {0.5, 1.0}, {v1, v2});
    const DupireLocalVol lv(srf);
    for (double k : ks) {
        const double v = lv.vol(k, 0.0);
        EXPECT_LE(v, DupireLocalVol::kCap);
        EXPECT_GE(v, DupireLocalVol::kFloor);
    }
    EXPECT_GT(lv.cap_count(), 0);
    EXPECT_DOUBLE_EQ(lv.vol(0.5, 0.0), DupireLocalVol::kCap);
    EXPECT_NEAR(lv.vol(0.0, 0.0), 0.05, 1e-9);
}

TEST(Dupire, CountersAreThreadSafe) {
    // MAJ-3: four threads hammering one shared object over a cap-firing
    // region must produce exactly the single-threaded count.
    std::vector<double> ks;
    std::vector<double> v1;
    for (int i = 0; i < 21; ++i) {
        const double k = -0.5 + 0.05 * i;
        ks.push_back(k);
        v1.push_back(0.05 + 4.0 * k * k);
    }
    std::vector<double> v2;
    for (double v : v1) v2.push_back(v * 1.05);
    const ImpliedVolSurface srf(ks, {0.5, 1.0}, {v1, v2});
    const DupireLocalVol lv(srf);
    const int per_thread = 100000;
    auto work = [&lv, per_thread]() {
        for (int i = 0; i < per_thread; ++i) {
            const double k = -0.5 + (i % 101) * 0.01;
            (void)lv.vol(k, 0.75);
        }
    };
    work();
    const long single_floor = lv.floor_count();
    const long single_cap = lv.cap_count();
    EXPECT_GT(single_cap, 0);
    lv.reset_counters();
    std::vector<std::thread> pool;
    for (int t = 0; t < 4; ++t) pool.emplace_back(work);
    for (auto& th : pool) th.join();
    EXPECT_EQ(lv.floor_count(), 4 * single_floor);
    EXPECT_EQ(lv.cap_count(), 4 * single_cap);
}
